// @wpackio/scripts/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "typelang", {
  enumerable: true,
  get: function () {
    return _babelConfig.typelang;
  }
});
Object.defineProperty(exports, "getBabelPresets", {
  enumerable: true,
  get: function () {
    return _babelConfig.getBabelPresets;
  }
});
Object.defineProperty(exports, "overrideBabelPresetOptions", {
  enumerable: true,
  get: function () {
    return _babelConfig.overrideBabelPresetOptions;
  }
});
Object.defineProperty(exports, "getDefaultBabelPresetOptions", {
  enumerable: true,
  get: function () {
    return _babelConfig.getDefaultBabelPresetOptions;
  }
});
Object.defineProperty(exports, "CreateWebpackConfig", {
  enumerable: true,
  get: function () {
    return _CreateWebpackConfig.CreateWebpackConfig;
  }
});
Object.defineProperty(exports, "WpackConfig", {
  enumerable: true,
  get: function () {
    return _CreateWebpackConfig.WpackConfig;
  }
});
Object.defineProperty(exports, "WebpackConfigHelper", {
  enumerable: true,
  get: function () {
    return _WebpackConfigHelper.WebpackConfigHelper;
  }
});
Object.defineProperty(exports, "WebpackConfigHelperConfig", {
  enumerable: true,
  get: function () {
    return _WebpackConfigHelper.WebpackConfigHelperConfig;
  }
});
Object.defineProperty(exports, "getProjectAndServerConfig", {
  enumerable: true,
  get: function () {
    return _getProjectAndServerConfig.getProjectAndServerConfig;
  }
});
Object.defineProperty(exports, "getProjectConfig", {
  enumerable: true,
  get: function () {
    return _getProjectAndServerConfig.getProjectConfig;
  }
});
Object.defineProperty(exports, "getServerConfig", {
  enumerable: true,
  get: function () {
    return _getProjectAndServerConfig.getServerConfig;
  }
});
Object.defineProperty(exports, "validateProjectConfig", {
  enumerable: true,
  get: function () {
    return _getProjectAndServerConfig.validateProjectConfig;
  }
});
Object.defineProperty(exports, "validateServerConfig", {
  enumerable: true,
  get: function () {
    return _getProjectAndServerConfig.validateServerConfig;
  }
});
Object.defineProperty(exports, "BannerConfig", {
  enumerable: true,
  get: function () {
    return _projectConfig.BannerConfig;
  }
});
Object.defineProperty(exports, "EntryConfig", {
  enumerable: true,
  get: function () {
    return _projectConfig.EntryConfig;
  }
});
Object.defineProperty(exports, "FileConfig", {
  enumerable: true,
  get: function () {
    return _projectConfig.FileConfig;
  }
});
Object.defineProperty(exports, "ProjectConfig", {
  enumerable: true,
  get: function () {
    return _projectConfig.ProjectConfig;
  }
});
Object.defineProperty(exports, "projectConfigDefault", {
  enumerable: true,
  get: function () {
    return _projectConfig.projectConfigDefault;
  }
});
Object.defineProperty(exports, "webpackLoaderOptionsOverride", {
  enumerable: true,
  get: function () {
    return _projectConfig.webpackLoaderOptionsOverride;
  }
});
Object.defineProperty(exports, "webpackOptionsOverrideFunction", {
  enumerable: true,
  get: function () {
    return _projectConfig.webpackOptionsOverrideFunction;
  }
});
Object.defineProperty(exports, "ServerConfig", {
  enumerable: true,
  get: function () {
    return _serverConfig.ServerConfig;
  }
});
Object.defineProperty(exports, "serverConfigDefault", {
  enumerable: true,
  get: function () {
    return _serverConfig.serverConfigDefault;
  }
});
Object.defineProperty(exports, "getFileLoaderForJsAndStyleAssets", {
  enumerable: true,
  get: function () {
    return _loaderHelpers.getFileLoaderForJsAndStyleAssets;
  }
});
Object.defineProperty(exports, "getFileLoaderOptions", {
  enumerable: true,
  get: function () {
    return _loaderHelpers.getFileLoaderOptions;
  }
});
Object.defineProperty(exports, "issuerForNonStyleFiles", {
  enumerable: true,
  get: function () {
    return _loaderHelpers.issuerForNonStyleFiles;
  }
});
Object.defineProperty(exports, "issuerForStyleFiles", {
  enumerable: true,
  get: function () {
    return _loaderHelpers.issuerForStyleFiles;
  }
});
Object.defineProperty(exports, "issuerForJsTsFiles", {
  enumerable: true,
  get: function () {
    return _loaderHelpers.issuerForJsTsFiles;
  }
});
Object.defineProperty(exports, "issuerForNonJsTsFiles", {
  enumerable: true,
  get: function () {
    return _loaderHelpers.issuerForNonJsTsFiles;
  }
});
Object.defineProperty(exports, "getResolvedLoader", {
  enumerable: true,
  get: function () {
    return _loaderHelpers.getResolvedLoader;
  }
});
Object.defineProperty(exports, "babelLoader", {
  enumerable: true,
  get: function () {
    return _loaderHelpers.babelLoader;
  }
});
Object.defineProperty(exports, "cssLoader", {
  enumerable: true,
  get: function () {
    return _loaderHelpers.cssLoader;
  }
});
Object.defineProperty(exports, "fileLoader", {
  enumerable: true,
  get: function () {
    return _loaderHelpers.fileLoader;
  }
});
Object.defineProperty(exports, "lessLoader", {
  enumerable: true,
  get: function () {
    return _loaderHelpers.lessLoader;
  }
});
Object.defineProperty(exports, "getStyleLoaderUses", {
  enumerable: true,
  get: function () {
    return _loaderHelpers.getStyleLoaderUses;
  }
});
Object.defineProperty(exports, "postCssLoader", {
  enumerable: true,
  get: function () {
    return _loaderHelpers.postCssLoader;
  }
});
Object.defineProperty(exports, "sassLoader", {
  enumerable: true,
  get: function () {
    return _loaderHelpers.sassLoader;
  }
});
Object.defineProperty(exports, "Build", {
  enumerable: true,
  get: function () {
    return _Build.Build;
  }
});
Object.defineProperty(exports, "Server", {
  enumerable: true,
  get: function () {
    return _Server.Server;
  }
});
Object.defineProperty(exports, "Bootstrap", {
  enumerable: true,
  get: function () {
    return _Bootstrap.Bootstrap;
  }
});
Object.defineProperty(exports, "ProjectDependencies", {
  enumerable: true,
  get: function () {
    return _Bootstrap.ProjectDependencies;
  }
});

var _babelConfig = require("./config/babelConfig");

var _CreateWebpackConfig = require("./config/CreateWebpackConfig");

var _WebpackConfigHelper = require("./config/WebpackConfigHelper");

var _getProjectAndServerConfig = require("./config/getProjectAndServerConfig");

var _projectConfig = require("./config/project.config.default");

var _serverConfig = require("./config/server.config.default");

var _loaderHelpers = require("./config/loaderHelpers");

var _Build = require("./scripts/Build");

var _Server = require("./scripts/Server");

var _Bootstrap = require("./scripts/Bootstrap");
