// @wpackio/scripts/lib/config/CreateWebpackConfig.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CreateWebpackConfig = void 0;

var _webpackMerge = require("webpack-merge");

var _projectConfig = require("./project.config.default");

var _serverConfig = require("./server.config.default");

var _WebpackConfigHelper = require("./WebpackConfigHelper");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * Create the final webpack config through this class.
 */
class CreateWebpackConfig {
  /**
   * Create an instance of GetWebpackConfig class.
   *
   * @param projectConfig Project configuration as recovered from user directory.
   * @param serverConfig Server configuration as recovered from user directory.
   * @param isDev Whether this is development mode.
   */
  constructor(projectConfig, serverConfig, cwd, isDev = true) {
    _defineProperty(this, "projectConfig", void 0);

    _defineProperty(this, "serverConfig", void 0);

    _defineProperty(this, "cwd", void 0);

    _defineProperty(this, "isDev", void 0);

    _defineProperty(this, "publicPath", void 0);

    // Create final configuration
    // By doing a shallow override
    this.projectConfig = _objectSpread(_objectSpread({}, _projectConfig.projectConfigDefault), projectConfig);
    this.serverConfig = _objectSpread(_objectSpread({}, _serverConfig.serverConfigDefault), serverConfig);
    this.cwd = cwd;
    this.isDev = isDev; // Also figure out the publicPath beforehand, because we do need it

    const {
      slug,
      outputPath,
      type
    } = this.projectConfig; // The public is generated automatically considering
    // the dev server is installed at root, and wp-content, themes, plugins
    // directories are all default.
    // But one can pass distPublicPath from serverConfig to override it.

    this.publicPath = this.serverConfig.distPublicPath || `/wp-content/${type}s/${slug}/${outputPath}/`;
  }
  /**
   * Get configuration object to feed to webpack.
   *
   * Depending on the project configuration, it could be for single compiler
   * as well as multi compiler.
   *
   * If `projectConfig.files` has length === 1, then it would be a single compiler
   * otherwise, it would be for multi compiler.
   */


  getWebpackConfig() {
    // Now it can be a single compiler, or multicompiler
    // In any case, figure it out, create the compiler options
    // and return the stuff.
    // If the configuration is for multiple compiler mode
    // Then return an array of config.
    if (this.isMultiCompiler()) {
      // Return an array of configuration
      const config = [];
      this.projectConfig.files.forEach(file => {
        config.push(this.getSingleWebpackConfig(file));
      });
      return config;
    } // Otherwise, just return a single compiler mode config


    return this.getSingleWebpackConfig(this.projectConfig.files[0]);
  }
  /**
   * Is the config going to be for multi-compiler?
   */


  isMultiCompiler() {
    return this.projectConfig.files.length > 1;
  }
  /**
   * Get devServer publicPath for all sorts of middlewares.
   */


  getPublicPath() {
    return this.publicPath;
  }
  /**
   * Get Hot Module Reload Path, which we get from WebpackConfigHelper
   * for consistency.
   *
   * This is tweaked so that react fast refresh works.
   */


  getHmrPath() {
    return _WebpackConfigHelper.WebpackConfigHelper.getHmrPath();
  }
  /**
   * Get Url to publicPath.
   */


  getPublicPathUrl() {
    return `${this.getServerUrl()}${this.publicPath}`;
  }
  /**
   * Get server URL where the hot server is live and waiting to become
   * awesome.
   */


  getServerUrl() {
    const {
      host,
      port
    } = this.serverConfig;
    return `//${host || 'localhost'}:${port}`;
  }
  /**
   * Get Webpack Configuration for single compiler mode.
   *
   * @param file Single file object.
   */


  getSingleWebpackConfig(file) {
    var _this$projectConfig$c, _this$projectConfig$c2, _this$projectConfig$c3, _this$projectConfig$c4;

    const {
      type,
      slug,
      hasReact,
      disableReactRefresh,
      hasSass,
      hasLess,
      hasFlow,
      bannerConfig,
      alias,
      optimizeSplitChunks,
      outputPath,
      appName,
      errorOverlay,
      externals,
      jsBabelOverride,
      jsBabelPresetOptions,
      tsBabelOverride,
      tsBabelPresetOptions,
      useReactJsxRuntime,
      disableWordPressExternals
    } = this.projectConfig; // get useBabelConfig from project, but override from file

    let useBabelConfig = this.projectConfig.useBabelConfig;

    if (file.useBabelConfig !== undefined) {
      useBabelConfig = file.useBabelConfig;
    } // figure out compile node modules


    const compileNodeModules = {
      dev: (_this$projectConfig$c = (_this$projectConfig$c2 = this.projectConfig.compileNodeModules) === null || _this$projectConfig$c2 === void 0 ? void 0 : _this$projectConfig$c2.dev) !== null && _this$projectConfig$c !== void 0 ? _this$projectConfig$c : true,
      prod: (_this$projectConfig$c3 = (_this$projectConfig$c4 = this.projectConfig.compileNodeModules) === null || _this$projectConfig$c4 === void 0 ? void 0 : _this$projectConfig$c4.prod) !== null && _this$projectConfig$c3 !== void 0 ? _this$projectConfig$c3 : true
    };
    const {
      host,
      port
    } = this.serverConfig;
    const helper = new _WebpackConfigHelper.WebpackConfigHelper(file, {
      appName,
      type,
      slug,
      host,
      port,
      hasReact,
      disableReactRefresh,
      hasSass,
      hasLess,
      hasFlow,
      bannerConfig,
      alias,
      optimizeSplitChunks,
      outputPath,
      publicPath: this.getPublicPath(),
      publicPathUrl: this.getPublicPathUrl(),
      errorOverlay,
      externals,
      useBabelConfig,
      jsBabelOverride,
      jsBabelPresetOptions,
      tsBabelOverride,
      tsBabelPresetOptions,
      useReactJsxRuntime,
      disableWordPressExternals,
      compileNodeModules
    }, this.cwd, this.isDev); // Now create the config and return it

    let config = _objectSpread({
      name: file.name,
      entry: helper.getEntry(),
      output: helper.getOutput(),
      module: helper.getModule(),
      plugins: helper.getPlugins(),
      resolve: helper.getResolve(),
      optimization: helper.getOptimization()
    }, helper.getCommon()); // Merge options if needed
    // Loose comparison because it could very well be undefined


    if (file.webpackConfig != null) {
      // If it is a function
      if (typeof file.webpackConfig === 'function') {
        config = file.webpackConfig(config, _webpackMerge.merge, helper.getAppDir(), this.isDev);
      } else {
        config = (0, _webpackMerge.merge)(config, file.webpackConfig);
      }
    }

    return config;
  }

}

exports.CreateWebpackConfig = CreateWebpackConfig;
