// @wpackio/scripts/lib/dev-utils/ops.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.fileExists = fileExists;
exports.hasTypeScript = hasTypeScript;
exports.wpackioRequestsToExternals = wpackioRequestsToExternals;
exports.defaultRequestToExternal = defaultRequestToExternal;
exports.wpackioRequestToHandle = wpackioRequestToHandle;
exports.defaultRequestToHandle = defaultRequestToHandle;
exports.basename = basename;
exports.getLockFilename = getLockFilename;
exports.lock = lock;
exports.unlock = unlock;
exports.BUNDLED_PACKAGES = exports.WORDPRESS_NAMESPACE = void 0;

var _fs = _interopRequireDefault(require("fs"));

var _path = _interopRequireDefault(require("path"));

var _lockfile = _interopRequireDefault(require("lockfile"));

var _util = _interopRequireDefault(require("util"));

var _os = _interopRequireDefault(require("os"));

var _crypto = _interopRequireDefault(require("crypto"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const lfUnlock = _util.default.promisify(_lockfile.default.unlock);
/**
 * Check if file exists or not using fs API.
 */


function fileExists(filepath) {
  try {
    // tslint:disable-next-line:non-literal-fs-path
    return _fs.default.statSync(filepath).isFile();
  } catch (_) {
    return false;
  }
}
/**
 * Check whether current working directory is a typescript project.
 *
 *
 * @param cwd Current working directory.
 * @returns True if tsconfig is found, false otherwiise.
 */


function hasTypeScript(cwd) {
  const tsconfigPath = _path.default.resolve(cwd, './tsconfig.json');

  return [fileExists(tsconfigPath), tsconfigPath];
}

const WORDPRESS_NAMESPACE = '@wordpress/';
exports.WORDPRESS_NAMESPACE = WORDPRESS_NAMESPACE;
const BUNDLED_PACKAGES = ['@wordpress/icons', '@wordpress/interface'];
/**
 * Given a string, returns a new string with dash separators converted to
 * camelCase equivalent. This is not as aggressive as `_.camelCase` in
 * converting to uppercase, where Lodash will also capitalize letters
 * following numbers.
 *
 * @see {https://github.com/WordPress/gutenberg/blob/c047e2716149c794eebff3c2c002f66a6f546f59/packages/dependency-extraction-webpack-plugin/lib/util.js#L83}
 *
 * @param {string} input Input dash-delimited string.
 * @return {string} Camel-cased string.
 */

exports.BUNDLED_PACKAGES = BUNDLED_PACKAGES;

function camelCaseDash(input) {
  return input.replace(/-([a-z])/g, (_, letter) => letter.toUpperCase());
}
/**
 * Wpackio specific request to global transformation. This is used in
 * @wordpress/dependency-extraction-webpack-plugin.
 *
 * Unlike the default one, it doesn't transform react/reactdom, moment, jQuery
 * etc.
 *
 * Transform @wordpress dependencies:
 * - request `@wordpress/api-fetch` becomes `[ 'wp', 'apiFetch' ]`
 * - request `@wordpress/i18n` becomes `[ 'wp', 'i18n' ]`
 *
 * @param request Module request (the module name in `import from`) to be transformed
 * @returns The resulting external definition. Return `undefined`
 *   to ignore the request. Return `string|string[]` to map the request to an external.
 */


function wpackioRequestsToExternals(request) {
  if (BUNDLED_PACKAGES.includes(request)) {
    return undefined;
  }

  if (request.startsWith(WORDPRESS_NAMESPACE)) {
    return ['wp', camelCaseDash(request.substring(WORDPRESS_NAMESPACE.length))];
  }

  return undefined;
}
/**
 * Default request to global transformation
 *
 * Transform @wordpress dependencies:
 * - request `@wordpress/api-fetch` becomes `[ 'wp', 'apiFetch' ]`
 * - request `@wordpress/i18n` becomes `[ 'wp', 'i18n' ]`
 *
 * @param {string} request Module request (the module name in `import from`) to be transformed
 * @return {string|string[]|undefined} The resulting external definition. Return `undefined`
 *   to ignore the request. Return `string|string[]` to map the request to an external.
 */


function defaultRequestToExternal(request) {
  switch (request) {
    case 'moment':
      return request;

    case '@babel/runtime/regenerator':
      return 'regeneratorRuntime';

    case 'lodash':
    case 'lodash-es':
      return 'lodash';

    case 'jquery':
      return 'jQuery';

    case 'react':
      return 'React';

    case 'react-dom':
      return 'ReactDOM';

    default:
      return wpackioRequestsToExternals(request);
  }
}
/**
 * Wpackio specific request to WordPress script handle transformation
 *
 * Transform @wordpress dependencies:
 * - request `@wordpress/i18n` becomes `wp-i18n`
 * - request `@wordpress/escape-html` becomes `wp-escape-html`
 *
 * @param {string} request Module request (the module name in `import from`) to be transformed
 * @return {string|undefined} WordPress script handle to map the request to. Return `undefined`
 *   to use the same name as the module.
 */


function wpackioRequestToHandle(request) {
  if (request.startsWith(WORDPRESS_NAMESPACE)) {
    return `wp-${request.substring(WORDPRESS_NAMESPACE.length)}`;
  }

  return undefined;
}
/**
 * Default request to WordPress script handle transformation
 *
 * Transform @wordpress dependencies:
 * - request `@wordpress/i18n` becomes `wp-i18n`
 * - request `@wordpress/escape-html` becomes `wp-escape-html`
 *
 * @param {string} request Module request (the module name in `import from`) to be transformed
 * @return {string|undefined} WordPress script handle to map the request to. Return `undefined`
 *   to use the same name as the module.
 */


function defaultRequestToHandle(request) {
  switch (request) {
    case '@babel/runtime/regenerator':
      return 'wp-polyfill';

    case 'lodash-es':
      return 'lodash';

    default:
      return wpackioRequestToHandle(request);
  }
}
/**
 * Get the name of the file from a file path.
 *
 * @param name Full path of the filename.
 * @returns Just the name of the file.
 */


function basename(name) {
  if (!name.includes('/')) {
    return name;
  }

  return name.substr(name.lastIndexOf('/') + 1);
}

function md5(data) {
  return _crypto.default.createHash('md5').update(data).digest('hex');
}
/**
 * Build a file path to a lock file in the tmp directory
 *
 * @param {string} filename
 */


function getLockFilename(filename) {
  const name = _path.default.basename(filename);

  const dirHash = md5(_path.default.dirname(filename));
  return _path.default.join(_os.default.tmpdir(), `${dirHash}-${name}.lock`);
}
/**
 * Create a lockfile (async)
 *
 * @param {string} filename
 */


async function lock(filename) {
  await new Promise((resolve, reject) => {
    _lockfile.default.lock(getLockFilename(filename), {
      wait: 6000,
      retryWait: 100,
      stale: 5000,
      retries: 100
    }, err => {
      if (err) {
        reject(err);
        return;
      }

      resolve();
    });
  });
}
/**
 * Remove a lockfile (async)
 *
 * @param {string} filename
 */


async function unlock(filename) {
  await lfUnlock(getLockFilename(filename));
}
