// @wpackio/scripts/lib/dev-utils/typescriptFormatter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.typescriptFormatter = typescriptFormatter;

var _os = _interopRequireDefault(require("os"));

var _codeFrame = require("@babel/code-frame");

var _chalk = _interopRequireDefault(require("chalk"));

var _fs = _interopRequireDefault(require("fs"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const issueOrigins = {
  typescript: 'TypeScript',
  internal: 'fork-ts-checker-webpack-plugin'
};

function typescriptFormatter(issue, cwd) {
  const {
    origin,
    severity,
    file,
    location,
    message,
    code
  } = issue;
  const colors = _chalk.default;
  const messageColor = severity === 'warning' ? colors.yellow : colors.red;
  const fileAndNumberColor = colors.bold.cyan;

  const source = file && _fs.default.existsSync(file) && _fs.default.readFileSync(file, 'utf-8');

  const frame = source ? (0, _codeFrame.codeFrameColumns)(source, location, {
    highlightCode: true,
    forceColor: true,
    linesAbove: 3,
    linesBelow: 1
  }).split('\n').map(str => `  ${str}`).join(_os.default.EOL) : '';
  return [messageColor.bold(`${issueOrigins[origin]} ${severity.toLowerCase()} in `) + fileAndNumberColor(`${file.replace(cwd, '.')}(${location.start.line},${location.start.column})`) + messageColor(':'), `${message}  ${messageColor.underline(`TS${code}`)}`, '', frame].join(_os.default.EOL);
}
